docs/designs/high_performance_ps.md

```diff
@@ -0,0 +1,59 @@
+# High Performance Parameter Server Design
+
+## Motivation
+
+This design doc focuses on implementing a high performance parameter server (PS). For the functionality of the PS, please refer to this [design doc](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/parameter_server.md)
+
+The PS receives gradients from workers, applies gradients to parameters, and sends the latest parameters to workers. Receiving gradients and sending parameters are primary I/O workloads of the PS, and parameters updating cost CPU resource. Since one PS could receive gradients from more than one worker, both I/O workload and CPU workload could be heavy.
+
+The current PS is in Python. Due to the existence of [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) of Python, gradients are applied to parameters sequentially with only one CPU core. As a result, the receiving gradients service is also blocked, and waiting for current gradients to be consumed. We want to remove this bottleneck and make full utilization of multiple CPU cores.
```

Comment:
Applying gradients to parameters should be sequentially to avoid overwriting the same parameters by different gradients in multiple threads. Tensorflow also use a lock in `Optimizer.apply_gradient` to update sequentially. I think we can transform serialized protobuf data to tensors parallelly but update parameters by those gradient tensors sequentially . Now, it is sequentially for both steps using Python.

Reply:
Yes, there are actually two different cases:

- If the embedding gradient is very sparse, the gradients could be updated in parallel. The gradients will be overwritten with a very small probability.

- If the gradient is dense, then, we should better apply it sequentially.

We plan to support both.